mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala

```diff
@@ -0,0 +1,243 @@
+      } else {
+        attrs = attrIterator.flatMap { head =>
+          attrs.map { tail =>
+            NumericAttribute.defaultAttr.withName(head.name.get + ":" + tail.name.get)
+          }
+        }
+      }
+    }
+    new AttributeGroup($(outputCol), attrs.toArray)
+  }
+
+  private def getAttributesAfterEncoding(
+      groupName: Option[String], attrs: Seq[Attribute]): Seq[Attribute] = {
+    def format(i: Int, attrName: Option[String], value: Option[String]): String = {
+      val parts = Seq(groupName, Some(attrName.getOrElse(i.toString)), value)
+      parts.flatten.mkString("_")
+    }
+    attrs.zipWithIndex.flatMap {
+      case (nominal: NominalAttribute, i) =>
+        if (nominal.values.isDefined) {
+          nominal.values.get.map(
+            v => BinaryAttribute.defaultAttr.withName(format(i, nominal.name, Some(v))))
+        } else {
+          Array.tabulate(nominal.getNumValues.get)(
+            j => BinaryAttribute.defaultAttr.withName(format(i, nominal.name, Some(j.toString))))
+        }
+      case (a: Attribute, i) =>
+        Seq(NumericAttribute.defaultAttr.withName(format(i, a.name, None)))
+    }
+  }
+
+  override def copy(extra: ParamMap): Interaction = defaultCopy(extra)
+
+  private def checkParams(): Unit = {
+    require(get(inputCols).isDefined, "Input cols must be defined first.")
+    require(get(outputCol).isDefined, "Output col must be defined first.")
+    require($(inputCols).length > 0, "Input cols must have non-zero length.")
+    require($(inputCols).distinct.length == $(inputCols).length, "Input cols must be distinct.")
+  }
+}
+
+/**  
+ * An iterator over VectorUDT or Double that one-hot encodes nominal fields in the output.
+ *
+ * @param cardinalities An array defining the cardinality of each vector sub-field, or a single
+ *                      value if the field is numeric. Fields with zero cardinality will not be
+ *                      one-hot encoded (output verbatim).
+ */
+// TODO(ekl) support drop-last option like OneHotEncoder does?
+private[ml] class EncodingIterator(cardinalities: Array[Int]) {
+  /** The size of the output vector. */
+  val size = cardinalities.map(i => if (i > 0) i else 1).sum
+
+  /**
+   * @param value The row to iterate over, either a Double or Vector.
+   * @param f The callback to invoke on each non-zero (index, value) output pair.
+   */
+  def foreachActive(value: Any, f: (Int, Double) => Unit) = value match {
+    case d: Double =>
+      assert(cardinalities.length == 1)
+      val numOutputCols = cardinalities(0)
+      if (numOutputCols > 0) {
+        assert(
+          d >= 0.0 && d == d.toInt && d < numOutputCols,
+          s"Values from column must be indices, but got $d.")
+        f(d.toInt, 1.0)
+      } else {
+        f(0, d)
+      }
+    case vec: Vector =>
+      assert(cardinalities.length == vec.size,
+        s"Vector column size was ${vec.size}, expected ${cardinalities.length}")
+      val dense = vec.toDense
```

Comment:
This is too expensive for sparse vectors. We should call `vec.foreachActive`. Storing the cumulative number of features could help.